storage-proofs/src/circuit/stacked/proof.rs

```diff
@@ -0,0 +1,603 @@
+use std::marker::PhantomData;
+
+use bellperson::{Circuit, ConstraintSystem, SynthesisError};
+use fil_sapling_crypto::circuit::{boolean::Boolean, num};
+use fil_sapling_crypto::jubjub::JubjubEngine;
+use paired::bls12_381::{Bls12, Fr};
+
+use crate::circuit::por::PoRCompound;
+use crate::circuit::{
+    constraint,
+    stacked::{hash::hash2, params::Proof},
+};
+use crate::compound_proof::{CircuitComponent, CompoundProof};
+use crate::drgraph::{Graph, BASE_DEGREE};
+use crate::hasher::Hasher;
+use crate::merklepor;
+use crate::parameter_cache::{CacheableParameters, ParameterSetMetadata};
+use crate::proof::ProofScheme;
+use crate::stacked::{StackedDrg, EXP_DEGREE};
+
+/// Stacked DRG based Proof of Replication.
+///
+/// # Fields
+///
+/// * `params` - parameters for the curve
+///
+pub struct StackedCircuit<'a, E: JubjubEngine, H: 'static + Hasher> {
+    params: &'a E::Params,
+    public_params: <StackedDrg<'a, H> as ProofScheme<'a>>::PublicParams,
+    replica_id: Option<H::Domain>,
+    comm_d: Option<H::Domain>,
+    comm_r: Option<H::Domain>,
+    comm_r_last: Option<H::Domain>,
+    comm_c: Option<H::Domain>,
+
+    // one proof per challenge
+    proofs: Vec<Proof<H>>,
+
+    _e: PhantomData<E>,
+}
+
+impl<'a, E: JubjubEngine, H: Hasher> CircuitComponent for StackedCircuit<'a, E, H> {
+    type ComponentPrivateInputs = ();
+}
+
+impl<'a, H: Hasher> StackedCircuit<'a, Bls12, H> {
+    #[allow(clippy::too_many_arguments)]
+    pub fn synthesize<CS>(
+        mut cs: CS,
+        params: &'a <Bls12 as JubjubEngine>::Params,
+        public_params: <StackedDrg<'a, H> as ProofScheme<'a>>::PublicParams,
+        replica_id: Option<H::Domain>,
+        comm_d: Option<H::Domain>,
+        comm_r: Option<H::Domain>,
+        comm_r_last: Option<H::Domain>,
+        comm_c: Option<H::Domain>,
+        proofs: Vec<Proof<H>>,
+    ) -> Result<(), SynthesisError>
+    where
+        CS: ConstraintSystem<Bls12>,
+    {
+        let circuit = StackedCircuit::<'a, Bls12, H> {
+            params,
+            public_params,
+            replica_id,
+            comm_d,
+            comm_r,
+            comm_r_last,
+            comm_c,
+            proofs,
+            _e: PhantomData,
+        };
+
+        circuit.synthesize(&mut cs)
+    }
+}
+
+impl<'a, H: Hasher> Circuit<Bls12> for StackedCircuit<'a, Bls12, H> {
+    fn synthesize<CS: ConstraintSystem<Bls12>>(self, cs: &mut CS) -> Result<(), SynthesisError> {
+        let StackedCircuit {
+            public_params,
+            proofs,
+            replica_id,
+            comm_r,
+            comm_d,
+            comm_r_last,
+            comm_c,
+            ..
+        } = self;
+
+        let graph = &public_params.graph;
+        let params = &self.params;
+        // In most cases (the exception being during testing) we want to ensure that the base and
+        // expansion degrees are the optimal values.
```

Comment:
By 'optimal', do you mean 'correct'?

Reply:
I did not write this

Reply:
I wrote this comment. The constant base and expansion degrees were an optimization brought in from r2 (I thought) and were not values that were ensuring correctness.

Reply:
Got it. Using global constants is itself the optimization. It's not that the specific values chosen are somehow more optimal. Rather, those will just be the ones required by the protocol and therefore 'correct' (anything else would be an incorrect implementation of the protocol). The check here is to make sure what we have hard-coded as an optimization agrees with what has been declared in the public parameters. In theory, we implement a range of possible proofs, varying by public parameters. In practice we are limited to performing only those which agree with our hard-coded constants.